### PR TITLE
Prototype web view of Tracer service information

### PR DESCRIPTION
**Pull request by Dr15Jones**
#### PR description:

Extended the edmTracerCompactLogViewer to output a data file that can be used by a crude web application to interactively view the tracer log information.

#### PR validation:

Viewed data generated from a Phase 2 reconstruction workflow and a trivial job with out of order finishing of LuminosityBlocks.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
please test

File: FWCore/Services/scripts/edmTracerCompactLogViewer.py [hunk 1]
```diff
@@ -203,24 +238,132 @@ def textPostfix(self):
...
     def textSpecial(self):
         return "starting"
+    def jsonInfo(self, counter, data):
+        if transitionIsGlobal(self.transition):
+            index = 0
+            if self.transition == Phase.startTracing:
+                data["globals"][0].append(jsonTransition(type=self.transition, id=index, sync=list(self.sync),start=0, finish=self.time ))
+                return
+            elif self.transition == Phase.esSync:
+                if self.sync[1] == 4294967295:
```
Comment (makortel) [inline FWCore/Services/scripts/edmTracerCompactLogViewer.py:278]: What is this magic value? Could make it a named constant or add a comment?

File: FWCore/Services/scripts/edmTracerCompactLogViewer.py [hunk 2]
```diff
@@ -203,24 +238,132 @@ def textPostfix(self):
...
+                while index+1 > len(globals):
+                    globals.append([])
+                container = globals[index]
+                #find source, should be previous
+                last = container[-1]
+                if last["type"]==Phase.globalBeginLumi and last["isSrc"]:
+                    last["sync"]=list(self.sync)
+                container.append(q[0])
+                container.append(q[1])
+            elif Phase.globalEndRun == self.transition or self.transition==Phase.globalBeginLumi or Phase.globalEndLumi == self.transition or self.transition==Phase.globalWriteRun or Phase.globalWriteLumi == self.transition:
```
Comment (makortel) [inline FWCore/Services/scripts/edmTracerCompactLogViewer.py:317]: What would you think about
```suggestion
            elif transition in [Phase.globalEndRun, Phase.globalBeginLumi, Phase.globalEndLumi, Phase.globalWriteRun, Phase.globalWriteLumi]:
```
?

File: FWCore/Services/scripts/edmTracerCompactLogViewer.py [hunk 3]
```diff
@@ -543,7 +906,24 @@ def textOutput( parser ):
     parser.add_argument('-f', '--frameworkOnly',
                         action='store_true',
                         help='''Output only the framework transitions, excluding the individual module transitions.''')
+    parser.add_argument('-j', '--json',
+                        action='store_true',
+                        help='''Write output in json format.''' )
+    parser.add_argument('-w', '--web',
+                        action='store_true',
+                        help='''Writes data.js file that can be used with the web based inspector. To use, copy directory ${CMSSW_RELEASE_BASE}/src/FWCore/Services/template/web to a web accessible area and move data.js into that directory.''')
```
Comment (makortel) [inline FWCore/Services/scripts/edmTracerCompactLogViewer.py:914]: How about adding a `README.md` that describes an example how to use the `Tracer` service and this script to create the `data.js` and use the web viewer?

File: FWCore/Services/scripts/edmTracerCompactLogViewer.py [hunk 4]
```diff
@@ -158,6 +167,32 @@ def transitionIndentLevel(transition):
...
+        return True
+    if transition == Phase.globalBeginRun or Phase.globalEndRun == transition:
+        return True
+    if transition == Phase.globalWriteRun:
+        return True
+    if transition == Phase.globalBeginLumi or Phase.globalEndLumi == transition:
+        return True
+    if transition == Phase.globalWriteLumi:
+        return True
+    if transition == Phase.esSyncEnqueue or transition == Phase.esSync:
```
[the remark this answers is not in the document]
Comment (Dr15Jones) [reply FWCore/Services/scripts/edmTracerCompactLogViewer.py]: done

File: FWCore/Services/scripts/edmTracerCompactLogViewer.py (hunk 2 above)
Comment (Dr15Jones) [reply FWCore/Services/scripts/edmTracerCompactLogViewer.py]: done

File: FWCore/Services/scripts/edmTracerCompactLogViewer.py (hunk 1 above)
Comment (Dr15Jones) [reply FWCore/Services/scripts/edmTracerCompactLogViewer.py]: The number is the largest allowed number for a LuminosityBlock (it is defined in C++). I've now used a named constant.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by Dr15Jones** [comment]
please test

**Comment by smuzaffar** [comment]
Milestone for this pull request has been moved to CMSSW_14_0_X.Please open a backport if it should also go in to CMSSW_13_3_X.

File: FWCore/Services/web/transitions.css [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  height: 100%;
+}
+
+.name_div {
+  width: 10%;
+  background-color: var(--colr-green);
+}
+#name_view {
+	flex:1;
+    width: 100%;
```
Comment (makortel) [inline FWCore/Services/web/transitions.css:47]: Indentation seems to be inconsistent in some places of this file (some lines use tabs and others use spaces)

File: FWCore/Services/web/transitions.js [hunk 6]
```diff
@@ -0,0 +1,496 @@
...
+	}
+	return maxTime;
+}
+
+const kEndTime = maxTime();
+
+function drawNames() {
+    name_context.save()
+    name_context.setTransform(1,0,0,1,0,0);
+	name_context.fillStyle = "#AACD6E"
```
Comment (makortel) [inline FWCore/Services/web/transitions.js:156]: Indentation seems to be inconsistent on some lines (some have tabs, others have spaces)

File: FWCore/Services/scripts/README.md
[the remark this answers is not in the document]
Comment (smuzaffar) [reply FWCore/Services/scripts/README.md:1]: yes , it is fine to have non-executable files in `scripts` directory.  `scram` only copies executable files from `scripts` to `cmssw/bin` PATH.

File: FWCore/Services/scripts/README.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
+#Using edmTracerCompactLogViewer.py
```
[the remark this answers is not in the document]
Comment (Dr15Jones) [reply FWCore/Services/scripts/README.md]: done

File: FWCore/Services/scripts/README.md [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+## Introduction
+The `edmTracerCompactLogViewer.py` is used to create a human understandable representation of the file generated by the `Tracer` service. Two representations are possible: text and interactive web.
+
+## Creating Tracer file
+To create a Tracer file, add the following to a cmsRun configuration
+```
```
[the remark this answers is not in the document]
Comment (Dr15Jones) [reply FWCore/Services/scripts/README.md]: done

File: FWCore/Services/scripts/README.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+The script will outut a new file named `data.js`.
+
+If you only want to see the framework transitions and not all the information about ED or ES modules, you can use the option `-f`.
+
+#### setup web data
+
+You will need to copy all the files in `$CMSSW_RELEASE_BASE/src/FWCore/Services/web` to a directory you can access from a web browser. Copy the `data.js` file
+created by `edmTracerCompactLogViewer.py` into that same web browser accessible directory. Now you can have your web browser access the `index.html` file.
+
+For directions on how to use the web application, see the `README.md` file in the `FWCore/Services/web` directory.
```
[the remark this answers is not in the document]
Comment (Dr15Jones) [reply FWCore/Services/scripts/README.md]: done

File: FWCore/Services/web/README.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
+# Interactive Tracer Log Viewer
+
+## Starting the viewer
+
+### standard web server
+If the output of edmTracerCompactLogViewer and the files from FWCore/Services/web have been placed in a directory accessible to a web server, you may just point your browser to the correct URL to run the viewer.
```
[the remark this answers is not in the document]
Comment (Dr15Jones) [reply FWCore/Services/web/README.md]: done

File: FWCore/Services/web/README.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+### local file
+By default, web browsers are not allowed to read files from the local disk. Many browsers give options to override that default. After overriding the default, make sure the output of edmTracerCompactLogViewer and the files from FWCore/Services/web have been placed in a directory on your local machine. Then tell  your web browser to open the index.html file in that directory to start the viewer.
```
[the remark this answers is not in the document]
Comment (Dr15Jones) [reply FWCore/Services/web/README.md]: done

File: FWCore/Services/web/transitions.css (hunk 5 above)
Comment (Dr15Jones) [reply FWCore/Services/web/transitions.css:47]: fixed indention

File: FWCore/Services/web/transitions.js (hunk 6 above)
Comment (Dr15Jones) [reply FWCore/Services/web/transitions.js]: reindented

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
+core

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
+core

**Comment by makortel** [comment]
Comparison differences are related to https://github.com/cms-sw/cmssw/issues/39803

**Comment by Dr15Jones** [comment]
please test

**Comment by makortel** [comment]
Comparison differences are related to https://github.com/cms-sw/cmssw/issues/39803

**Comment by makortel** [comment]
+core

**Comment by makortel** [comment]
@cms-sw/orp-l2 Could you merge this PR? Thanks!

**Comment by antoniovilela** [comment]
+1

**Pull request merged**